certora/specs/Reverts.spec

```diff
@@ -0,0 +1,253 @@
+
+    // Ensure that no interest is accumulated.
+    require mLastUpdate == e.block.timestamp;
+
+    uint256 collateralPrice = mockPrice();
+
+    (_, pBorrowShares, pCollateral) = MORPHO.position(currentContract.ID, borrower);
+
+    mathint collateralQuoted = require_uint256(summaryMulDivDown(pCollateral, collateralPrice, ORACLE_SCALE()));
+
+    // Safe require because the implementation would revert, see rule zeroCollateralQuotedReverts.
+    require collateralQuoted > 0;
+
+    mathint borrowed = require_uint256(summaryToAssetsUp(pBorrowShares, mTotalBorrowAssets, mTotalBorrowShares));
+    mathint ltv = require_uint256(wDivUp(require_uint256(borrowed), require_uint256(collateralQuoted)));
+
+    preLiquidate@withrevert(e, borrower, seizedAssets, 0, data);
+
+    // Ensure the price is unchanged in the preLiquidate call.
+    require !priceChanged;
+
+    assert ltv <= currentContract.PRE_LLTV => lastReverted;
+}
+
+// Check that pre-liqudidating a position such that ltv > LLTV would revert.
+rule liquidatablePositionReverts(env e,address borrower, uint256 seizedAssets, bytes data) {
+    // Market values.
+    uint256 mTotalBorrowAssets;
+    uint256 mTotalBorrowShares;
+    uint256 mLastUpdate;
+
+    // Position values.
+    uint256 pBorrowShares;
+    uint256 pCollateral;
+
+    requireInvariant preLltvConsistent();
+    requireInvariant preLCFConsistent();
+    requireInvariant preLIFConsistent();
+
+    (_, _, mTotalBorrowAssets,mTotalBorrowShares,mLastUpdate, _) = MORPHO.market(currentContract.ID);
+
+    // Ensure that no interest is accumulated.
+    require mLastUpdate == e.block.timestamp;
+
+    uint256 collateralPrice = mockPrice();
+
+    (_, pBorrowShares, pCollateral) = MORPHO.position(currentContract.ID, borrower);
+
+    mathint collateralQuoted = require_uint256(summaryMulDivDown(pCollateral, collateralPrice, ORACLE_SCALE()));
+
+    // Safe require because the implementation would revert, see rule zeroCollateralQuotedReverts.
+    require collateralQuoted > 0;
+
+    mathint borrowed = require_uint256(summaryToAssetsUp(pBorrowShares, mTotalBorrowAssets, mTotalBorrowShares));
+    mathint ltv = require_uint256(wDivUp(require_uint256(borrowed), require_uint256(collateralQuoted)));
+
+    preLiquidate@withrevert(e, borrower, seizedAssets, 0, data);
+
+    // Ensure the price is unchanged in the preLiquidate call.
+    require !priceChanged;
+
+    assert  ltv > currentContract.LLTV => lastReverted;
+}
+
+rule excessivePreliquidationReverts(env e,address borrower, uint256 seizedAssets, bytes data) {
+    uint256 mTotalBorrowAssets;
+    uint256 mTotalBorrowShares;
+    uint256 mLastUpdate;
+
+    uint256 pBorrowShares;
+    uint256 pCollateral;
+
+    (_, _, mTotalBorrowAssets,mTotalBorrowShares,mLastUpdate, _) = MORPHO.market(currentContract.ID);
+
+    // Ensure that no interest is accumulated.
+    require mLastUpdate == e.block.timestamp;
```

Comment:
Note that it is fine to assume that `lastUpdate(marketParams().id()) == e.block.timestamp` because `preLiquidate` calls `accrueInterest(marketParams())` and that changes `lastUpdate(marketParams().id())` to be `e.block.timestamp`. But here we assume that `lastUpdate(ID) == e.block.timestamp` which implicitely assumes that the invariant `ID == marketParams().id()` holds. Maybe it's fine though, if we prove it somewhere else

Reply:
You mean this? https://github.com/morpho-org/pre-liquidation/blob/bf7933fa8ac8a93deaf5c22b5794988fdd053981/certora/specs/Health.spec#L86

Reply:
Let's implement this one, because now we know that this market lastUpdate is non zero (so `hasOfMarketParams` precondition is true for the market of this pre-liquidation contract)

Reply:
Implemented in https://github.com/morpho-org/pre-liquidation/pull/37/commits/350020475fe843e54d61676820b7ce55ecd6eff2.